logverification/app/appentry.go

```diff
@@ -0,0 +1,292 @@
+type MMREntryFields struct {
+
+	// Domain defines the hashing schema for the MMR Entry
+	Domain byte
+
+	// SerializedBytes are app (customer) provided fields in the MMR Entry, serialized in a consistent way.
+	SerializedBytes []byte
+}
+
+// AppEntry is the app provided data for a corresponding log entry.
+//
+// It contains key information for verifying inclusion of the corresponding log entry.
+//
+// NOTE: all fields are sourced from the app data, or derived from it.
+// NONE of the fields in an AppEntry are sourced from the log.
+type AppEntry struct {
+	// AppId is an identifier of the app committing the merkle log entry
+	AppId string
+
+	// LogId is a uuid in byte form of the specific log identifier
+	LogId []byte
+
+	// ExtraBytes are extrabytes provided by datatrails for the specific app
+	ExtraBytes []byte
+
+	// MMREntryFields used to determine the MMR Entry
+	MMREntryFields *MMREntryFields
+
+	// MerkleLogCommit used to define information about the log entry
+	MerkleLogCommit *assets.MerkleLogCommit
+
+	// MerkleLogConfirm used to define information about the log seal
+	MerkleLogConfirm *assets.MerkleLogConfirm
+}
+
+// NewAppEntry creates a new app entry entry
+func NewAppEntry(
+	appId string,
+	logId []byte,
+	extraBytes []byte,
+	mmrEntryFields *MMREntryFields,
+	merklelogCommit *assets.MerkleLogCommit,
+) *AppEntry {
+
+	verifiableLogEntry := &AppEntry{
+		AppId:           appId,
+		LogId:           logId,
+		ExtraBytes:      extraBytes,
+		MMREntryFields:  mmrEntryFields,
+		MerkleLogCommit: merklelogCommit,
+	}
+
+	return verifiableLogEntry
+}
+
+// MMREntry derives the mmr entry of the corresponding log entry from the app data.
+//
+// MMREntry is:
+//   - H( Domain | MMR Salt | Serialized Bytes)
+func (ae *AppEntry) MMREntry() ([]byte, error) {
+
+	hasher := sha256.New()
+
+	// domain
+	hasher.Write([]byte{ae.MMREntryFields.Domain})
+
+	// mmr salt
+	mmrSalt, err := ae.MMRSalt()
+	if err != nil {
+		return nil, err
+	}
+
+	hasher.Write(mmrSalt)
+
+	// serialized bytes
+	hasher.Write(ae.MMREntryFields.SerializedBytes)
+
+	return hasher.Sum(nil), nil
+
+}
+
+// MMRIndex gets the mmr index of the corresponding log entry.
+func (ae *AppEntry) MMRIndex() uint64 {
+	return ae.MerkleLogCommit.Index
+}
+
+// MMRSalt derives the MMR Salt of the corresponding log entry from the app data.
+// MMRSalt is the datatrails provided fields included on the MMR Entry.
+//
+// this is (extrabytes | idtimestamp) for any apps that adhere to log entry version 1.
+func (ve *AppEntry) MMRSalt() ([]byte, error) {
+
+	mmrSalt := make([]byte, MMRSaltSize)
+
+	copy(mmrSalt[:24], ve.ExtraBytes)
+
+	// get the byte representation of idtimestamp
+	idTimestamp, _, err := massifs.SplitIDTimestampHex(ve.MerkleLogCommit.Idtimestamp)
+	if err != nil {
+		return nil, err
+	}
+
+	idTimestampBytes := make([]byte, IDTimestapSizeBytes)
+	binary.BigEndian.PutUint64(idTimestampBytes, idTimestamp)
+
+	copy(mmrSalt[24:], idTimestampBytes)
```

Comment:
Its probably worth testing the boundary case here of extra bytes that run up to the boundary to guard against accidental breakage in future

Reply:
done!